### PR TITLE
Okta Identity Engine | Changes to Okta hosted sign in code to fix social login and sessions

### PR DESCRIPTION
**Pull request by coldlink**
## What does this change?

Upgrading to Okta Identity Engine on CODE broke a few things in our code environment, mainly due to how we do the [interception](https://github.com/guardian/gateway/pull/1699) on the Okta hosted sign in page.

The tl;dr to as to why this interception is required it to be able to customise the login experience with a completely Guardian branded sign in flow tailored to our needs, and at the time the hosted page didn't support many of the features that we need, e.g. registration, reset password.

For now, before we investigate whether Okta can now provide the features we need, this PR fixes the issues that arised with the OIE upgrade.

### Social Sign In

The first thing that wasn't working was social sign in. We noticed that when clicking on the social sigin button it would redirect back to the sign in page without taking the user to the social provider to authenticate.

This is because Okta changed the way that Social IDP authentication worked, instead of using HTTP redirects, Okta now use JavaScript redirects and new endpoints in order to authenticate with social, as mentioned here: https://developer.okta.com/docs/guides/custom-widget/main/#hide-or-suppress-the-transient-sign-in-widget

So we have to rewrite the Okta interception code to show the sign in widget which now handles social sign in, if there is a call to social sign in, specifically if the `idp` query parameter is present. We also have to check for the `'/oauth2/v1/authorize/callback'` which is the callback used to get back from the social provider to Okta once authentication is completed via social.

https://github.com/guardian/gateway/blob/02aff8198cc263e406323f767786e66e7266fa8a/scripts/okta/okta-login.ts#L36-L49

If these parameters are not detected we continue with the interception as we did before.

We also had to add the `/idp` and `/sso` paths to fastly, which we did in fastly here https://github.com/guardian/identity-platform/pull/682, and have added it to our cypress nginx config too for CI.

This works, except we still see that the widget is visible when doing social sign in.

Using similar code as explained in https://developer.okta.com/docs/guides/custom-widget/main/#hide-or-suppress-the-transient-sign-in-widget, we can add some variables and CSS to hide the widget when social sign in occurs.

https://github.com/guardian/gateway/blob/02aff8198cc263e406323f767786e66e7266fa8a/scripts/okta/okta-login.html#L14-L19

Other changes we made was to update the `login-default.html` file which is the default html that Okta provide for the hosted login page, unfortunately there doesn't seem to be an option in the dashboard to revert the page to default html, so we add it to this repo, now updated for Okta Identity engine.

We also refactor `getRedirectUrl` to now take in a `URLSearchParams` object instead of the direct query string, as we can share the params between the social sign in and the interception code.

### `idx` cookie

In Okta Identity Engine there is a new main authentication cookie, the `idx` cookie which replaces the `sid` cookie from Identity Classic.

If Okta detects a `sid` cookie, it will set an `idx` cookie, but will no longer refresh the `sid` cookie, though it will still be valid.

In https://github.com/guardian/gateway/pull/2422 we updated Gateway to be able to support the `idx` cookie alongside the `sid` cookie, however after upgrading to CODE, we found the session refreshing test to break, as the `sid` cookie is no longer refreshed.

So we update that one specific test to check for the `idx` cookie instead of the `sid` cookie.

There were other quirks for the `idx` cookie mentioned in https://github.com/guardian/identity-platform/pull/682

All the other cypress ete tests in Gateway should be updated to also check for the `idx` cookie, but we'll do this in a separate PR to avoid polluting this one with the essential changes, as the `sid` cookie still works and matches what is closer to PROD.

## Other changes

We've split the Cypress Okta End to End tests for registration into 2 files.

From

`registration.2.cy.ts`

To

`registration_1.2.cy.ts`
`registration_2.6.cy.ts`

This allows us to parallelise those tests, as they were taking long to run when in a single file and was timing out after 15 mins.

By splitting them into 2 files, and adding another machine to the e2e tests should mean it runs much faster.

Documentation on how this works is in our `setup.md` documentation

https://github.com/guardian/gateway/blob/b02b034c735614bdd575d14eef1b4a2b200027ca/docs/setup.md?plain=1#L234-L244

### CODE REVIEW

File: scripts/okta/okta-login.ts [hunk 1]
```diff
@@ -30,11 +30,36 @@ import { getRedirectUrl } from './lib/helper';
...
-	window.OktaUtil,
-);
-
-window.location.replace(redirectUrl);
+const locationSearch = window.location.search;
+const searchParams = new URLSearchParams(locationSearch);
+
+if (
+	searchParams.has('idp') ||
+	window.location.pathname === '/oauth2/v1/authorize/callback'
```
Comment (kelvin-chappell) [inline scripts/okta/okta-login.ts:38]: Perhaps this condition needs some more explanation or a link to an explanation

File: cypress/integration/ete-okta/sign_in.1.cy.ts [hunk 2]
```diff
@@ -129,12 +128,13 @@ describe('Sign in flow, Okta enabled', () => {
...
-							expect(newSidCookie).to.exist;
-							expect(newSidCookie?.value).to.equal(originalSidCookie?.value);
-							if (newSidCookie?.expiry && originalSidCookie?.expiry) {
-								expect(newSidCookie?.expiry).to.be.greaterThan(
-									originalSidCookie?.expiry,
+						cy.getCookie('idx').then((newIdxCookie) => {
+							expect(newIdxCookie).to.exist;
+							// `idx` cookie doesn't have same value as original when refreshed, which is different to the Okta Classic `sid` cookie
+							expect(newIdxCookie?.value).to.not.equal(orignalIdxCookie?.value);
+							if (newIdxCookie?.expiry && orignalIdxCookie?.expiry) {
```
[the remark this answers is not in the document]
Comment (coldlink) [reply cypress/integration/ete-okta/sign_in.1.cy.ts:135]: This was to get typescript to pass compilation. In Cypress types the type for `Cookie.expiry` is optional, so could be `undefined`, we have to check that the values are defined, and then do the check to see that they're not the same

[the remark this answers is not in the document]
Comment (coldlink) [reply cypress/integration/ete-okta/sign_in.1.cy.ts:135]: I suppose I should add a cypress check to see if the expiry value exists, as the only other time they might be `undefined` is if a cookie is set as a session cookie, and we don't want that!

File: scripts/okta/okta-login.ts (hunk 1 above)
Comment (coldlink) [reply scripts/okta/okta-login.ts:43]: Added in a comment above the if statement!

File: cypress/integration/ete-okta/sign_in.1.cy.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (coldlink) [reply cypress/integration/ete-okta/sign_in.1.cy.ts:139]: Updated to explicitly check if the expiry exists as a cypress test.

**Review by kelvin-chappell — DISMISSED** [review]
👍

**Review by kelvin-chappell — DISMISSED** [review]

**Review by kelvin-chappell — APPROVED** [review]

**Pull request merged**